timemachine/cpp/src/segmented_weighted_random_sampler.hpp

```diff
@@ -0,0 +1,54 @@
+#pragma once
+
+#include <vector>
+
+#include "curand.h"
+#include "device_buffer.hpp"
+#include <cub/util_type.cuh>
+
+namespace timemachine {
+
+// A SegmentedWeightedRandomSampler for random sampling from batches of probabilities (np.random.choice(X, replace=True p=probabilities))
+// Uses a similar approach as the segmented APIs in cub
+template <typename RealType> class SegmentedWeightedRandomSampler {
+
+private:
+    const int max_vals_per_segment_; // Max number of values that can be had per segment
+    const int num_segments_;         // Number of segments to run at a time
+    size_t temp_storage_bytes_;
+
+    // Stores both the initial uniform random values and the final gumbel distribution
+    DeviceBuffer<RealType> d_gumbel_;
+    DeviceBuffer<cub::KeyValuePair<int, RealType>> d_arg_max_;
+
+    DeviceBuffer<char> d_argmax_storage_;
+
+    curandGenerator_t cr_rng_;
+
+public:
+    SegmentedWeightedRandomSampler(const int max_vals_per_segment, const int segments, const int seed);
+
+    ~SegmentedWeightedRandomSampler();
+
+    void sample_device(
+        const int total_values,
+        const int num_segments,
+        const int *d_segments,
```

Comment:
naming nit: maybe `d_segment_offsets`? I was initially confused about what this represents.

Reply:
Good call, renamed in https://github.com/proteneer/timemachine/pull/1235/commits/683d8018e359f0c204abc6c400d6a6cb67fbb790